coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/functions/HiveInOperator.java

```diff
@@ -73,46 +76,42 @@ public void unparse(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec)
     writer.endList(listFrame);
   }
 
-  // copied from Calcite' SqlInOperator
+  // Adapts the parent's behavior to allow IN to have any number of arguments (i.e., list elements).
   public RelDataType deriveType(SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
     final List<SqlNode> operands = call.getOperandList();
-    assert operands.size() == 2;
     final SqlNode left = operands.get(0);
-    final SqlNode right = operands.get(1);
+    final SqlNodeList right = new SqlNodeList(ZERO);
+    for (int i = 1; i < operands.size(); i++) {
+      right.add(operands.get(i));
+    }
 
     final RelDataTypeFactory typeFactory = validator.getTypeFactory();
     RelDataType leftType = validator.deriveType(scope, left);
     RelDataType rightType;
 
     // Derive type for RHS.
-    if (right instanceof SqlNodeList) {
-      // Handle the 'IN (expr, ...)' form.
-      List<RelDataType> rightTypeList = new ArrayList<>();
-      SqlNodeList nodeList = (SqlNodeList) right;
-      for (int i = 0; i < nodeList.size(); i++) {
-        SqlNode node = nodeList.get(i);
-        RelDataType nodeType = validator.deriveType(scope, node);
-        rightTypeList.add(nodeType);
-      }
-      rightType = typeFactory.leastRestrictive(rightTypeList);
-
-      // First check that the expressions in the IN list are compatible
-      // with each other. Same rules as the VALUES operator (per
-      // SQL:2003 Part 2 Section 8.4, <in predicate>).
-      if (null == rightType && validator.isTypeCoercionEnabled()) {
-        // Do implicit type cast if it is allowed to.
-        rightType = validator.getTypeCoercion().getWiderTypeFor(rightTypeList, true);
-      }
-      if (null == rightType) {
-        throw validator.newValidationError(right, RESOURCE.incompatibleTypesInList());
-      }
-
-      // Record the RHS type for use by SqlToRelConverter.
-      ((SqlValidatorImpl) validator).setValidatedNodeType(nodeList, rightType);
-    } else {
```

Comment:
I believe it's because `right` is now explicitly built as a `SqlNodeList` using the list of operands, so there is no case where `right` isn't an instance of `SqlNodeList`.

https://github.com/linkedin/coral/pull/410/files#diff-0e69daa34dc3fe4e131718bf050d0cfd47dde9e79921e53e6ccda1d78aeefe17R83-R86